kubernetes/config/authn_plugin/plugins/oidc/libkubernetes_oidc.c

```diff
@@ -0,0 +1,318 @@
+#include "authn_plugin.h"
+#include "authn_plugin_util.h"
+#include "kube_config_util.h"
+#include "kube_config_common.h"
+#include "binary.h"
+#include "cJSON.h"
+#include <time.h>
+#include <errno.h>
+
+#define OIDC_ID_TOKEN_DELIM "."
+#define OIDC_ID_TOKEN_EXP "exp"
+#define OIDC_CONFIGURATION_URL_TEMPLATE "%s/.well-known/openid-configuration"
+#define OIDC_TOKEN_ENDPOINT "token_endpoint"
+#define OIDC_ID_TOKEN "id_token"
+#define OIDC_REFRESH_TOKEN "refresh_token"
+
+#define JWT_PART2_BUFFER_SIZE 1024
+#define BASE64_PADDING_STRING_SIZE 4
+#define BASE64_PADDING_CHAR "="
+
+#define REFRESH_TOKEN_CONTENT_TYPE "application/x-www-form-urlencoded"
+#define REFRESH_TOKEN_CREDENTIAL_TEMPLATE "%s:%s"
+#define REFRESH_TOKEN_POST_DATA_TEMPLATE "refresh_token=%s&grant_type=refresh_token"
+
+#define OIDC_CONFIGURATION_URL_BUFFER_SIZE 1024
+#define REFRESH_TOKEN_CREDENTIAL_BUFFER_SIZE 1024
+#define REFRESH_TOKEN_POST_DATA_BUFFER_SIZE 1024
+
+static time_t get_token_expiration_time(const char *token_string)
+{
+    static char fname[] = "get_token_expiration_time()";
+
+    time_t expiration_time = 0;
+
+    if (!token_string || '\0' == token_string[0]) {
+        return 0;
+    }
+    char *dup_token_string = strdup(token_string);
+    if (!dup_token_string) {
+        return 0;
+    }
+
+    char *p = NULL;
+    p = strtok(dup_token_string, OIDC_ID_TOKEN_DELIM);  /* jwt header */
+    if (!p) {
+        fprintf(stderr, "%s: The token <%s> is not a valid JWT token.\n", fname, token_string);
+        goto end;
+    }
+    p = strtok(NULL, OIDC_ID_TOKEN_DELIM);  /* jwt part2 */
+    if (!p) {
+        fprintf(stderr, "%s: The token <%s> is not a valid JWT token.\n", fname, token_string);
+        goto end;
+    }
+
+    int base64_padding_length = 4 - strlen(p) % 4;
+    char base64_padding_string[BASE64_PADDING_STRING_SIZE];
+    memset(base64_padding_string, 0, sizeof(base64_padding_string));
+    for (int i = 0; i < base64_padding_length; i++) {
+        strncat(base64_padding_string, BASE64_PADDING_CHAR, 1);
```

Comment:
given that you just `memset` everything to zero, why do you need to strncat the padding?

Reply:
This code snippets is used to generate base64 padding string for "base64_padding_string" ( the content is "=" , "==",  or "==="  )

for the first time (i = 1),
```strncat``` actually works same as ```strncpy```  
"" (empty string) --> “=”

but the second time (i = 2) or the third time (i=3)
```strncat``` will do the string concatenation.
"="  --> "=="
"=="  --> "==="